internal/hooks/v0hooks/v0hooks.go

```diff
@@ -1,98 +1,97 @@
 package v0hooks
 
 import (
+	"net/http"
+	"time"
+
 	"github.com/gofrs/uuid"
 	"github.com/golang-jwt/jwt/v5"
 	"github.com/supabase/auth/internal/mailer"
 	"github.com/supabase/auth/internal/models"
+	"github.com/supabase/auth/internal/utilities"
 )
 
-type HookType string
+type Name string
 
 const (
-	PostgresHook HookType = "pg-functions"
+	SendSMS              Name = "send-sms"
+	SendEmail            Name = "send-email"
+	CustomizeAccessToken Name = "customize-access-token"
+	MFAVerification      Name = "mfa-verification"
+	PasswordVerification Name = "password-verification"
+	BeforeUserCreated    Name = "before-user-created"
+	AfterUserCreated     Name = "after-user-created"
 )
 
-// Hook Names
 const (
 	HookRejection = "reject"
 )
 
-type HTTPHookInput interface {
-	IsHTTPHook()
+const (
+	DefaultMFAHookRejectionMessage      = "Further MFA verification attempts will be rejected."
+	DefaultPasswordHookRejectionMessage = "Further password verification attempts will be rejected."
+)
+
+type Header struct {
+	UUID uuid.UUID `json:"uuid"`
+	Time time.Time `json:"time"`
+
+	// Hook name
+	Name Name `json:"name,omitempty"`
+
+	// IP Address of the request, if present
+	IPAddress string `json:"ip_address,omitempty"`
+}
+
+func NewHeader(r *http.Request, name Name) *Header {
+	return &Header{
+		UUID:      uuid.Must(uuid.NewV4()),
+		Time:      time.Now(),
+		IPAddress: utilities.GetIPAddress(r),
+		Name:      name,
+	}
 }
 
-type HookOutput interface {
-	IsError() bool
-	Error() string
-	GetHookError() AuthHookError
+type BeforeUserCreatedRequest struct {
+	Header *Header      `json:"header"`
```

Comment:
Would prefer the concept of "header" be renamed to "metadata" as header immediately forces you (as a junior developer just seeing Auth hooks) to think of HTTP Headers which don't make sense if you're implementing the hook as a Postgres function. Furthermore this implementation of the concept encodes metadata, not HTTP headers.

Reply:
Sure, done